Framework/Algorithms/src/PolarizationCorrections/PolarizerEfficiency.cpp

```diff
@@ -0,0 +1,185 @@
+#include "MantidKernel/BoundedValidator.h"
+#include "MantidKernel/CompositeValidator.h"
+#include "MantidKernel/ListValidator.h"
+
+#include <boost/algorithm/string/join.hpp>
+#include <filesystem>
+
+namespace Mantid::Algorithms {
+// Register the algorithm into the algorithm factory
+DECLARE_ALGORITHM(PolarizerEfficiency)
+
+using namespace Kernel;
+using namespace API;
+
+namespace PropertyNames {
+static const std::string INPUT_WORKSPACE = "InputWorkspace";
+static const std::string ANALYSER_EFFICIENCY = "AnalyserEfficiency";
+static const std::string SPIN_STATES = "SpinStates";
+static const std::string OUTPUT_WORKSPACE = "OutputWorkspace";
+static const std::string OUTPUT_FILE_PATH = "OutputFilePath";
+} // namespace PropertyNames
+
+void PolarizerEfficiency::init() {
+  // Declare required input parameters for algorithm and do some validation here
+  auto validator = std::make_shared<CompositeValidator>();
+  validator->add<WorkspaceUnitValidator>("Wavelength");
+  validator->add<HistogramValidator>();
+  declareProperty(
+      std::make_unique<WorkspaceProperty<>>(PropertyNames::INPUT_WORKSPACE, "", Direction::Input, validator),
+      "Input group workspace to use for polarization calculation");
+  const auto &wavelengthValidator = std::make_shared<WorkspaceUnitValidator>("Wavelength");
+  declareProperty(std::make_unique<WorkspaceProperty<MatrixWorkspace>>(PropertyNames::ANALYSER_EFFICIENCY, "",
+                                                                       Direction::Input, wavelengthValidator),
+                  "Analyser efficiency as a function of wavelength");
+  declareProperty(std::make_unique<WorkspaceProperty<MatrixWorkspace>>(PropertyNames::OUTPUT_WORKSPACE, "",
+                                                                       Direction::Output, PropertyMode::Optional),
+                  "Polarizer efficiency as a function of wavelength");
+
+  const auto &spinValidator = std::make_shared<SpinStateValidator>(std::unordered_set<int>{4});
+  declareProperty(PropertyNames::SPIN_STATES, "11,10,01,00", spinValidator,
+                  "Order of individual spin states in the input group workspace, e.g. \"01,11,00,10\"");
+
+  declareProperty(std::make_unique<FileProperty>(PropertyNames::OUTPUT_FILE_PATH, "", FileProperty::OptionalSave),
+                  "File name or path for the output to be saved to.");
+}
+
+/**
+ * Tests that the inputs are all valid
+ * @return A map containing the incorrect workspace
+ * properties and an error message
+ */
+std::map<std::string, std::string> PolarizerEfficiency::validateInputs() {
+  std::map<std::string, std::string> errorList;
+  const std::string inputWorkspaceName = getProperty(PropertyNames::INPUT_WORKSPACE);
+  if (!AnalysisDataService::Instance().doesExist(inputWorkspaceName)) {
+    errorList[PropertyNames::INPUT_WORKSPACE] = "The workspace " + inputWorkspaceName + " does not exist in the ADS.";
+    return errorList;
+  }
```

Comment:
I don't think the workspace needs to be in the ADS for the algorithm to work. If it's going to be used as a child algorithm (like we plan to do with the workflow algorithm to cover the whole correction process) then the input workspaces are likely going to not be in the ADS. You can get the workspace directly from the input by doing something like this:
```suggestion
  const WorkspaceGroup_sptr inputWorkspace = getProperty(PropertyNames::INPUT_WORKSPACE);
  if (inputWorkspace == nullptr) {
    errorList[PropertyNames::INPUT_WORKSPACE] = "The input workspace  is not a workspace group."
    return errorList;
  }
```